File: momentra/momentra/Shell/GroupActive/Shared/GroupInvitePeopleSheet.swift
import SwiftUI
import UIKit

/// Invite link/QR + active members manage (organizer role edit / remove).
struct GroupInvitePeopleSheet: View {
    let momentId: String
    let momentTitle: String
    let momentTypeCode: String
    let currentUserId: String?
    @Binding var isPresented: Bool
    var onSaved: () -> Void = {}

    @State private var inviteCode: String?
    @State private var minting = true
    @State private var mintError: String?
    @State private var participants: [APIClient.GroupParticipantPayload] = []
    @State private var actionError: String?
    @State private var busyParticipantId: String?
    @State private var removeTarget: APIClient.GroupParticipantPayload?
    @State private var copied = false
    @State private var guestName = ""
    @State private var addingGuest = false

    private let sheetBg = TripSheetTokens.bg
    private let roleOptions: [(code: String, label: String)] = [
        ("PARTICIPANT", "Member"),
        ("ORGANIZER", "Organizer"),
        ("OBSERVER", "Viewer"),
    ]

    private var displayPath: String? {
        inviteCode.map { GroupInviteLink.displayPath(code: $0) }
    }

    private var copyText: String? {
        inviteCode.map { GroupInviteLink.copyText(code: $0) }
    }

    private var qrPayload: String? {
        inviteCode.map { GroupInviteLink.qrPayload(code: $0) }
    }

    private var activeMembers: [APIClient.GroupParticipantPayload] {
        participants.filter { ($0.status ?? "").uppercased() == "ACTIVE" }
    }

    private var viewerIsOrganizer: Bool {
        guard let currentUserId else { return false }
        return activeMembers.contains { p in
            p.userId == currentUserId &&
            ["ORGANIZER", "CO_ORGANIZER"].contains((p.roleCode ?? "").uppercased())
        }
    }

    var body: some View {
        NativeSheetScaffold(
            title: "Invite People",
            onClose: { isPresented = false },
            background: sheetBg
        ) {
            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    TripSheetHeader(
                        iconAsset: "GroupQaUserPlus",
                        title: "Invite People",
                        subtitle: "Share a link or QR so people can join",
                        accent: TripForm.accent
                    )

                    TripInviteShareSection(
                        minting: minting,
                        displayPath: displayPath,
                        mintError: mintError,
                        copied: $copied,
                        copyText: copyText,
                        qrPayload: qrPayload,
                        momentTitle: momentTitle
                    )

                    if viewerIsOrganizer {
                        VStack(alignment: .leading, spacing: 8) {
                            TripFieldLabel(text: "Add guest")
                            Text("Guests have no account — organizers can add them to expenses.")
                                .font(.plusJakarta(size: 11))
                                .foregroundStyle(TripForm.muted)
                            HStack(spacing: 8) {
                                TextField("Guest name", text: $guestName)
                                    .font(.plusJakarta(size: 13))
                                    .foregroundStyle(TripForm.text)
                                    .padding(10)
                                    .background(TripForm.field)
                                    .overlay(RoundedRectangle(cornerRadius: 8).stroke(TripForm.border))
                                    .clipShape(RoundedRectangle(cornerRadius: 8))
                                    .accessibilityIdentifier("group.invite.guestName")
                                Button {
                                    Task { await addGuest() }
                                } label: {
                                    Text(addingGuest ? "…" : "Add")
                                        .font(.plusJakarta(size: 13, weight: .semibold))
                                        .foregroundStyle(.white)
                                        .padding(.horizontal, 14)
                                        .padding(.vertical, 10)
                                        .background(TripForm.accent)
                                        .clipShape(RoundedRectangle(cornerRadius: 8))
                                }
                                .buttonStyle(.plain)
                                .disabled(addingGuest || guestName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                                .accessibilityIdentifier("group.invite.addGuest")
                            }
                        }
                    }

                    if !activeMembers.isEmpty {
                        VStack(alignment: .leading, spacing: 8) {
                            TripFieldLabel(text: "Active members (\(activeMembers.count))")
                            ForEach(activeMembers) { p in
                                memberRow(p)
                            }
                        }
                    }

                    if let actionError {
                        Text(actionError)
                            .font(.plusJakarta(size: 12))
                            .foregroundStyle(Color(hex: "#F87171"))
                    }

                    Text("Share via Messages or WhatsApp so they can join")
                        .font(.plusJakarta(size: 12))
                        .foregroundStyle(TripForm.muted)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
                .padding(24)
            }
        } footer: {
            EmptyView()
        }
        .presentationDetents([.large])
        .task(id: "\(momentId)|\(momentTitle)|\(momentTypeCode)") {
            await bootstrap()
        }
        .confirmationDialog(
            "Remove member?",
            isPresented: Binding(
                get: { removeTarget != nil },
                set: { if !$0 { removeTarget = nil } }
            ),
            titleVisibility: .visible
        ) {
            Button("Remove", role: .destructive) {
                if let target = removeTarget {
                    removeTarget = nil
                    Task { await removeMember(target) }
                }
            }
            Button("Cancel", role: .cancel) { removeTarget = nil }
        } message: {
            Text("Remove \(removeTarget?.displayName ?? "this member") from the group?")
        }
    }

    @ViewBuilder
    private func memberRow(_ p: APIClient.GroupParticipantPayload) -> some View {
        let name = (p.displayName?.isEmpty == false) ? (p.displayName ?? "Member") : "Member"
        let busy = busyParticipantId == p.participantId
        let selected = uiRoleCode(p.roleCode)
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 10) {
                Circle()
                    .fill(TripForm.accent.opacity(0.2))
                    .frame(width: 28, height: 28)
                    .overlay(
                        Text(tripInitials(name))
                            .font(.plusJakarta(size: 10, weight: .bold))
                            .foregroundStyle(TripForm.accent)
                    )
                VStack(alignment: .leading, spacing: 2) {
                    HStack(spacing: 6) {
                        Text(name)
                            .font(.plusJakarta(size: 13, weight: .semibold))
                            .foregroundStyle(TripForm.text)
                        if p.guest {
                            Text("Guest")
                                .font(.plusJakarta(size: 10, weight: .semibold))
                                .foregroundStyle(TripForm.accent)
                                .padding(.horizontal, 6)
                                .padding(.vertical, 2)
                                .background(TripForm.accent.opacity(0.15))
                                .clipShape(Capsule())
                        }
                    }
                    Text(p.roleLabel ?? displayRoleLabel(p.roleCode))
                        .font(.plusJakarta(size: 11))
                        .foregroundStyle(TripForm.muted)
                }
                Spacer()
                Text("Active")
                    .font(.plusJakarta(size: 11, weight: .semibold))
                    .foregroundStyle(TripForm.green)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(TripForm.green.opacity(0.12))
                    .clipShape(Capsule())
                if viewerIsOrganizer {
                    Button {
                        removeTarget = p
                    } label: {
                        Text("Remove")
                            .font(.plusJakarta(size: 12, weight: .semibold))
                            .foregroundStyle(busy ? TripForm.muted : Color(hex: "#EF4444"))
                            .padding(.horizontal, 10)
                            .padding(.vertical, 6)
                            .background(Color(hex: "#EF4444").opacity(0.12))
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                    .buttonStyle(.plain)
                    .disabled(busy)
                    .accessibilityLabel("Remove \(p.displayName ?? "member")")
                }
            }
            if viewerIsOrganizer && !p.guest {
                HStack(spacing: 8) {
                    ForEach(roleOptions, id: \.code) { opt in
                        let isSelected = selected == opt.code
                        Text(opt.label)
                            .font(.plusJakarta(size: 11, weight: .semibold))
                            .foregroundStyle(isSelected ? TripForm.text : TripForm.muted)
                            .padding(.horizontal, 12)
                            .padding(.vertical, 6)
                            .background(isSelected ? TripForm.accent.opacity(0.25) : TripForm.field)
                            .overlay(
                                Capsule().stroke(isSelected ? TripForm.accent : TripForm.border)
                            )
                            .clipShape(Capsule())
                            .onTapGesture {
                                guard !busy, !isSelected else { return }
                                Task { await changeRole(p, to: opt.code) }
                            }
                    }
                }
            }
        }
        .padding(10)
        .background(TripForm.field)
        .overlay(RoundedRectangle(cornerRadius: 8).stroke(TripForm.border))
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }

    private func displayRoleLabel(_ roleCode: String?) -> String {
        switch (roleCode ?? "").uppercased() {
        case "ORGANIZER", "CO_ORGANIZER": return "Organizer"
        case "OBSERVER", "VIEWER": return "Viewer"
        case "RESIDENT": return "Resident"
        case "CONTRIBUTOR": return "Contributor"
        default: return "Member"
        }
    }

    private func uiRoleCode(_ roleCode: String?) -> String {
        switch (roleCode ?? "").uppercased() {
        case "ORGANIZER", "CO_ORGANIZER": return "ORGANIZER"
        case "OBSERVER", "VIEWER": return "OBSERVER"
        default: return "PARTICIPANT"
        }
    }

    private func bootstrap() async {
        inviteCode = nil
        minting = true
        mintError = nil
        do {
            let invite = try await APIClient.shared.mintGroupInvite(
                title: momentTitle.isEmpty ? "Trip" : momentTitle,
                momentTypeCode: momentTypeCode.isEmpty ? "TRIP" : momentTypeCode,
                momentId: momentId,
                idempotencyKey: UUID().uuidString
            )
            inviteCode = invite.inviteCode
            minting = false
        } catch {
            mintError = error.localizedDescription
            minting = false
        }
        await refreshParticipants()
    }

    private func refreshParticipants() async {
        do {
            participants = try await APIClient.shared.listGroupParticipants(momentId: momentId)
        } catch {
            // best-effort list
        }
    }

    private func addGuest() async {
        let name = guestName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return }
        addingGuest = true
        actionError = nil
        defer { addingGuest = false }
        do {
            _ = try await APIClient.shared.addGroupParticipant(momentId: momentId, displayName: name)
            guestName = ""
            await refreshParticipants()
            onSaved()
        } catch {
            actionError = error.localizedDescription
        }
    }

    private func changeRole(_ p: APIClient.GroupParticipantPayload, to roleCode: String) async {
        busyParticipantId = p.participantId
        actionError = nil
        do {
            _ = try await APIClient.shared.updateGroupParticipantRole(
                momentId: momentId,
                participantId: p.participantId,
                roleCode: roleCode
            )
            await refreshParticipants()
            onSaved()
        } catch {
            actionError = error.localizedDescription
        }
        busyParticipantId = nil
    }

    private func removeMember(_ p: APIClient.GroupParticipantPayload) async {
        busyParticipantId = p.participantId
        actionError = nil
        do {
            _ = try await APIClient.shared.removeGroupParticipant(
                momentId: momentId,
                participantId: p.participantId
            )
            await refreshParticipants()
            onSaved()
        } catch {
            actionError = error.localizedDescription
        }
        busyParticipantId = nil
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/group/shared/GroupInvitePeopleSheet.kt
package com.example.momentra.ui.shell.group.shared

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.momentra.R
import com.example.momentra.data.api.GroupParticipantDto
import com.example.momentra.data.repository.GroupSliceRepository
import com.example.momentra.ui.shell.empty.group.GroupInviteLink
import com.example.momentra.ui.shell.empty.group.generateInviteQrBitmap
import com.example.momentra.ui.shell.empty.group.inviteMessage
import com.example.momentra.ui.shell.empty.group.sendInviteSms
import com.example.momentra.ui.shell.empty.group.sendInviteWhatsApp
import com.example.momentra.ui.shell.empty.group.shareInviteLink
import com.example.momentra.ui.shell.empty.group.shareInviteQr
import com.example.momentra.ui.theme.PlusJakartaSans
import kotlinx.coroutines.launch
import java.util.Locale

private val MANAGE_ROLE_OPTIONS = listOf(
    "PARTICIPANT" to "Member",
    "ORGANIZER" to "Organizer",
    "OBSERVER" to "Viewer",
)

private fun displayRoleLabel(roleCode: String?, isGuest: Boolean = false): String {
    if (isGuest) return "Guest"
    return when (roleCode?.uppercase(Locale.US)) {
        "ORGANIZER", "CO_ORGANIZER" -> "Organizer"
        "OBSERVER", "VIEWER" -> "Viewer"
        "RESIDENT" -> "Resident"
        "CONTRIBUTOR" -> "Contributor"
        else -> "Member"
    }
}

private fun uiRoleCode(roleCode: String?): String = when (roleCode?.uppercase(Locale.US)) {
    "ORGANIZER", "CO_ORGANIZER" -> "ORGANIZER"
    "OBSERVER", "VIEWER" -> "OBSERVER"
    else -> "PARTICIPANT"
}

/** Invite link/QR + active members manage (organizer role edit / remove). */
@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@Composable
fun GroupInvitePeopleSheet(
    momentId: String,
    momentTitle: String,
    momentTypeCode: String,
    visible: Boolean,
    onDismiss: () -> Unit,
    onSaved: () -> Unit = {},
    currentUserId: String? = null,
    repository: GroupSliceRepository = remember { GroupSliceRepository() },
) {
    if (!visible) return
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    var inviteCode by remember { mutableStateOf<String?>(null) }
    var minting by remember { mutableStateOf(true) }
    var mintError by remember { mutableStateOf<String?>(null) }
    var participants by remember { mutableStateOf<List<GroupParticipantDto>>(emptyList()) }
    var actionError by remember { mutableStateOf<String?>(null) }
    var busyParticipantId by remember { mutableStateOf<String?>(null) }
    var removeTarget by remember { mutableStateOf<GroupParticipantDto?>(null) }
    var copied by remember { mutableStateOf(false) }
    var guestName by remember { mutableStateOf("") }
    var addingGuest by remember { mutableStateOf(false) }

    fun refreshParticipants() {
        scope.launch {
            repository.getParticipants(momentId).fold(
                onSuccess = { participants = it.participants },
                onFailure = { /* list is best-effort */ },
            )
        }
    }

    LaunchedEffect(momentId, momentTitle, momentTypeCode) {
        inviteCode = null
        minting = true
        mintError = null
        repository.mintInviteForMoment(
            title = momentTitle.ifBlank { "Trip" },
            momentTypeCode = momentTypeCode.ifBlank { "TRIP" },
            momentId = momentId,
        ).fold(
            onSuccess = {
                inviteCode = it.inviteCode
                minting = false
            },
            onFailure = {
                mintError = it.message ?: "Could not mint invite"
                minting = false
            },
        )
        refreshParticipants()
    }

    val displayPath = inviteCode?.let { GroupInviteLink.displayPath(it) }
    val copyText = inviteCode?.let { GroupInviteLink.copyText(it) }
    val qrPayload = inviteCode?.let { GroupInviteLink.qrPayload(it) }
    val qrSizePx = with(LocalDensity.current) { 144.dp.roundToPx() }
    val qrBitmap = remember(qrPayload, qrSizePx) {
        qrPayload?.let { generateInviteQrBitmap(it, qrSizePx) }?.asImageBitmap()
    }
    val activeMembers = participants.filter { it.status.equals("ACTIVE", ignoreCase = true) }
    val viewerIsOrganizer = activeMembers.any { p ->
        p.userId != null &&
            p.userId == currentUserId &&
            p.roleCode.uppercase(Locale.US) in setOf("ORGANIZER", "CO_ORGANIZER")
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = TripSheet.Bg,
        dragHandle = {
            Box(
                modifier = Modifier
                    .padding(top = 12.dp, bottom = 4.dp)
                    .size(width = 40.dp, height = 5.dp)
                    .clip(RoundedCornerShape(100.dp))
                    .background(Color.White.copy(alpha = 0.2f)),
            )
        },
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .navigationBarsPadding()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 24.dp)
                .padding(bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            TripSheetHeaderRow(
                title = "Invite People",
                subtitle = "Share a link or QR so people can join",
                iconRes = R.drawable.ic_group_qa_userplus,
                accent = TripSheetTokens.Accent,
            )

            TripInviteShareSection(
                minting = minting,
                displayPath = displayPath,
                mintError = mintError,
                copied = copied,
                copyText = copyText,
                qrBitmap = qrBitmap,
                onCopy = {
                    val text = copyText ?: return@TripInviteShareSection
                    val cm = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                    cm.setPrimaryClip(ClipData.newPlainText("invite", text))
                    copied = true
                },
                onShareLink = {
                    val text = copyText ?: return@TripInviteShareSection
                    shareInviteLink(context, text)
                },
                onShareQr = {
                    val text = copyText ?: return@TripInviteShareSection
                    val bitmap = qrPayload?.let { generateInviteQrBitmap(it, 512) } ?: return@TripInviteShareSection
                    shareInviteQr(context, bitmap, text)
                },
                onMessages = {
                    val text = copyText ?: return@TripInviteShareSection
                    sendInviteSms(context, null, inviteMessage(momentTitle, text))
                },
                onWhatsApp = {
                    val text = copyText ?: return@TripInviteShareSection
                    sendInviteWhatsApp(context, null, inviteMessage(momentTitle, text))
                },
            )

            if (viewerIsOrganizer) {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    TripFieldLabel("Add guest")
                    Text(
                        "Guests have no account — organizers can add them to expenses.",
                        color = TripSheetTokens.Muted,
                        fontSize = 11.sp,
                        fontFamily = PlusJakartaSans,
                    )
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        BasicTextField(
                            value = guestName,
                            onValueChange = { guestName = it },
                            singleLine = true,
                            textStyle = LocalTextStyle.current.copy(
                                color = TripSheetTokens.Text,
                                fontSize = 13.sp,
                                fontFamily = PlusJakartaSans,
                            ),
                            cursorBrush = SolidColor(TripSheetTokens.Accent),
                            modifier = Modifier
                                .weight(1f)
                                .clip(RoundedCornerShape(8.dp))
                                .background(TripSheetTokens.Field)
                                .border(1.dp, TripSheetTokens.Border, RoundedCornerShape(8.dp))
                                .padding(horizontal = 10.dp, vertical = 12.dp)
                                .testTag("group.invite.guestName"),
                            decorationBox = { inner ->
                                if (guestName.isEmpty()) {
                                    Text(
                                        "Guest name",
                                        color = TripSheetTokens.Muted,
                                        fontSize = 13.sp,
                                        fontFamily = PlusJakartaSans,
                                    )
                                }
                                inner()
                            },
                        )
                        Text(
                            if (addingGuest) "…" else "Add",
                            color = Color.White,
                            fontSize = 13.sp,
                            fontWeight = FontWeight.SemiBold,
                            fontFamily = PlusJakartaSans,
                            modifier = Modifier
                                .clip(RoundedCornerShape(8.dp))
                                .background(TripSheetTokens.Accent)
                                .clickable(
                                    enabled = !addingGuest && guestName.trim().isNotEmpty(),
                                ) {
                                    val name = guestName.trim()
                                    if (name.isEmpty()) return@clickable
                                    scope.launch {
                                        addingGuest = true
                                        actionError = null
                                        repository.addParticipant(momentId, name).fold(
                                            onSuccess = {
                                                guestName = ""
                                                addingGuest = false
                                                refreshParticipants()
                                                onSaved()
                                            },
                                            onFailure = {
                                                addingGuest = false
                                                actionError = it.message
                                            },
                                        )
                                    }
                                }
                                .padding(horizontal = 14.dp, vertical = 10.dp)
                                .testTag("group.invite.addGuest"),
                        )
                    }
                }
            }

            if (activeMembers.isNotEmpty()) {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    TripFieldLabel("Active members (${activeMembers.size})")
                    activeMembers.forEach { p ->
                        InviteActiveMemberRow(
                            participant = p,
                            viewerIsOrganizer = viewerIsOrganizer,
                            busy = busyParticipantId == p.participantId,
                            onRoleChange = { nextRole ->
                                scope.launch {
                                    busyParticipantId = p.participantId
                                    actionError = null
                                    repository.updateParticipantRole(
                                        momentId = momentId,
                                        participantId = p.participantId,
                                        roleCode = nextRole,
                                    ).fold(
                                        onSuccess = {
                                            busyParticipantId = null
                                            refreshParticipants()
                                            onSaved()
                                        },
                                        onFailure = {
                                            busyParticipantId = null
                                            actionError = it.message
                                        },
                                    )
                                }
                            },
                            onRemove = { removeTarget = p },
                        )
                    }
                }
            }

            actionError?.let {
                Text(it, color = Color(0xFFF87171), fontSize = 12.sp, fontFamily = PlusJakartaSans)
            }

            Text(
                "Share via Messages or WhatsApp so they can join",
                color = TripSheetTokens.Muted,
                fontSize = 12.sp,
                fontFamily = PlusJakartaSans,
                modifier = Modifier.fillMaxWidth(),
            )
        }
    }

    removeTarget?.let { target ->
        AlertDialog(
            onDismissRequest = { removeTarget = null },
            title = { Text("Remove member?", fontFamily = PlusJakartaSans, fontWeight = FontWeight.Bold) },
            text = {
                Text(
                    "Remove ${target.displayName?.takeIf { it.isNotBlank() } ?: "this member"} from the group?",
                    fontFamily = PlusJakartaSans,
                    fontSize = 14.sp,
                )
            },
            confirmButton = {
                TextButton(
                    onClick = {
                        val p = target
                        removeTarget = null
                        scope.launch {
                            busyParticipantId = p.participantId
                            actionError = null
                            repository.removeParticipant(momentId, p.participantId).fold(
                                onSuccess = {
                                    busyParticipantId = null
                                    refreshParticipants()
                                    onSaved()
                                },
                                onFailure = {
                                    busyParticipantId = null
                                    actionError = it.message
                                },
                            )
                        }
                    },
                ) {
                    Text("Remove", color = Color(0xFFEF4444), fontWeight = FontWeight.Bold)
                }
            },
            dismissButton = {
                TextButton(onClick = { removeTarget = null }) {
                    Text("Cancel", color = TripSheetTokens.Muted)
                }
            },
        )
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun InviteActiveMemberRow(
    participant: GroupParticipantDto,
    viewerIsOrganizer: Boolean,
    busy: Boolean,
    onRoleChange: (String) -> Unit,
    onRemove: () -> Unit,
) {
    val name = participant.displayName?.takeIf { it.isNotBlank() } ?: "Member"
    val selectedUiRole = uiRoleCode(participant.roleCode)
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(TripSheetTokens.Field)
            .border(1.dp, TripSheetTokens.Border, RoundedCornerShape(8.dp))
            .padding(10.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            Box(
                modifier = Modifier
                    .size(28.dp)
                    .clip(CircleShape)
                    .background(TripSheetTokens.Accent.copy(alpha = 0.2f)),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    name.trim().firstOrNull()?.uppercaseChar()?.toString() ?: "?",
                    color = TripSheetTokens.Accent,
                    fontSize = 10.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = PlusJakartaSans,
                )
            }
            Column(modifier = Modifier.weight(1f)) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Text(name, color = TripSheetTokens.Text, fontWeight = FontWeight.SemiBold, fontSize = 13.sp, fontFamily = PlusJakartaSans)
                    if (participant.isGuest) {
                        Text(
                            "Guest",
                            color = TripSheetTokens.Accent,
                            fontSize = 10.sp,
                            fontWeight = FontWeight.SemiBold,
                            fontFamily = PlusJakartaSans,
                            modifier = Modifier
                                .clip(RoundedCornerShape(999.dp))
                                .background(TripSheetTokens.Accent.copy(alpha = 0.15f))
                                .padding(horizontal = 6.dp, vertical = 2.dp),
                        )
                    }
                }
                Text(
                    participant.roleLabel ?: displayRoleLabel(participant.roleCode, participant.isGuest),
                    color = TripSheetTokens.Muted,
                    fontSize = 11.sp,
                    fontFamily = PlusJakartaSans,
                )
            }
            Text(
                "Active",
                color = TripFormTokens.Green,
                fontSize = 11.sp,
                fontWeight = FontWeight.SemiBold,
                fontFamily = PlusJakartaSans,
                modifier = Modifier
                    .clip(RoundedCornerShape(999.dp))
                    .background(TripFormTokens.Green.copy(alpha = 0.12f))
                    .padding(horizontal = 8.dp, vertical = 4.dp),
            )
            if (viewerIsOrganizer) {
                Icon(
                    Icons.Filled.Close,
                    contentDescription = "Remove",
                    tint = if (busy) TripSheetTokens.Muted else Color(0xFFEF4444),
                    modifier = Modifier
                        .size(18.dp)
                        .then(if (busy) Modifier else Modifier.clickable(onClick = onRemove)),
                )
            }
        }
        if (viewerIsOrganizer && !participant.isGuest) {
            FlowRow(
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                MANAGE_ROLE_OPTIONS.forEach { (code, label) ->
                    val selected = selectedUiRole == code
                    Text(
                        label,
                        color = if (selected) TripSheetTokens.Text else TripSheetTokens.Muted,
                        fontSize = 11.sp,
                        fontWeight = FontWeight.SemiBold,
                        fontFamily = PlusJakartaSans,
                        modifier = Modifier
                            .clip(RoundedCornerShape(999.dp))
                            .background(if (selected) TripSheetTokens.Accent else TripSheetTokens.Bg)
                            .border(
                                1.dp,
                                if (selected) TripSheetTokens.Accent else TripSheetTokens.Border,
                                RoundedCornerShape(999.dp),
                            )
                            .then(
                                if (!busy && !selected) {
                                    Modifier.clickable { onRoleChange(code) }
                                } else {
                                    Modifier
                                },
                            )
                            .padding(horizontal = 12.dp, vertical = 6.dp),
                    )
                }
            }
        }
    }
}
